include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,374 @@
+      rescheduleOpSt_;
+  std::optional<unifex::any_scheduler> sched_;
+  // Keep track of the consumer scheduler
+  unifex::any_scheduler consumerSched_{_default_scheduler};
+  std::exception_ptr exception_;
+  // In this case, this keeps the consumer coroutine + a done() continuation.
+  // it's needed for at_coroutine exit for now, but also whenw e handle stop
+  // requests.
+  unifex::continuation_handle<> continuation_;
+  std::optional<any_receiver_ref<T&>> receiverOpt_;
+};
+
+}  // namespace detail
+
+template <typename T>
+class [[nodiscard]] async_generator {
+public:
+  using promise_type = detail::async_generator_promise<T>;
+
+  async_generator() noexcept : coroutine_(nullptr) {}
+
+  explicit async_generator(promise_type& promise) noexcept
+    : coroutine_(
+          unifex::coro::coroutine_handle<promise_type>::from_promise(promise)) {
+  }
+
+  async_generator(async_generator&& other) noexcept
+    : coroutine_(other.coroutine_) {
+    other.coroutine_ = nullptr;
+  }
+
+  ~async_generator() {
+    if (coroutine_) {
+      coroutine_.destroy();
+    }
+  }
+
+  async_generator& operator=(async_generator&& other) noexcept {
+    async_generator temp(std::move(other));
+    swap(temp);
+    return *this;
+  }
+
+  async_generator(const async_generator&) = delete;
+  async_generator& operator=(const async_generator&) = delete;
+
+  // Potential problem: not sure if get_scheduler(gen.next()) would return
+  // the right thing here. Perhaps we need a wrapper sender that also records
+  // each next sender's scheduler and customizes get_scheduler(...)
+  auto next() noexcept {
+    // grab the receiver's scheduler; assume that the next-sender
+    // is always started on the corresponding context
+    return detail::read_scheduler() | let_value([this](auto sched) {
+             if (!coroutine_.promise().sched_) {
+               // capture the receiver's scheduler as the stream's
+               // scheduler on the first run of the next-sender
+               coroutine_.promise().sched_ = sched;
+             }
+             coroutine_.promise().consumerSched_ = sched;
+
+             // check to see if we're currently running on the saved scheduler
+             return just_void_or_done(coroutine_.promise().sched_ == sched)
+                 // get back on the desired scheduler when we're not already
+                 // there
+                 | let_done([this]() {
+                      return schedule(*coroutine_.promise().sched_);
+                    }) |
+                 let_value([this]() {
+                      // once we're on the right scheduler, use create<>()
+                      // to resume the generator's coroutine_handle<> after
+                      // saving the create-receiver in the promise so we can
+                      // complete create-sender from within the generator
+                      return create<T>([this](auto& rec) {
```

Comment:
I think you should use `create<T&>` here. The values generated by this generator will all be passed to the next-receiver using code like this:
```
co_yield <expr>;
```
so the lifetime of `<expr>` will span the suspend point created by the `co_yield`. You can therefore efficiently and safely pass `<expr>` as an lvalue reference through to the next-receiver.
```suggestion
                      return create<T&>([this](auto& rec) {
```